Objects/unicodeobject.c

```diff
@@ -3835,7 +3835,7 @@ static PyObject*
 unicode_decode_locale(const char *str, Py_ssize_t len,
                       _Py_error_handler errors, int current_locale)
 {
-    if (str[len] != '\0' || (size_t)len != strlen(str))  {
+    if (str[len] != '\0' || (size_t)len != strnlen(str, len))  {
```

Comment:
Good catch. I agree that there is a bug and that `strnlen()` should be used.

Would be nice to add a test, but this may be not easy.

Reply:
I guess adress sanitizer test could catch that.

Reply:
Both of these conditions can be written as

     (size_t)len != strnlen(str, len + 1)

right?

Reply:
The +1 is not necessary.

Reply:
It isn't if you have the first disjunct. But I said both of them can be replaced by what I've written. Or do you think that _both_ of them can be replaced by what I've said without +1?

Reply:
Ah ok, sounds about right :)

Reply:
At second look, the current code is correct. `strlen()` never reads past the buffer because of the first condition. It is guaranteed that `strlen()` does not read more than `len+1` bytes.

Therefore, there is no bug and the code change is not necessary.

Reply:
Yes, but if it took you 2.5 months to see that, I'd say the code is not really clearest. :-) I never said there's a bug, but that it can be written more clearly if we're allowed to use strnlen. But if there's a strict "don't fix it if it ain't broken" policy, I understand.

Reply:
The bug was actually in an earlier version of cpython. Anyway, can't hurt to clean up the code.